src/porepy/models/metric.py

```diff
@@ -0,0 +1,257 @@
+
+"""
+
+from functools import partial
+from typing import Callable
+
+import numpy as np
+
+import porepy as pp
+
+
+class EuclideanMetric:
+    """Plain Euclidean norm for variables and equations."""
+
+    def _euclidean_norm(self, values: np.ndarray) -> float:
+        """Compute the Euclidean norm of an array.
+
+        Parameters:
+            values: array to compute the norm of.
+
+        Returns:
+            float: measure of values
+
+        """
+        return np.linalg.norm(values) / np.sqrt(values.size)
+
+    def variable_norm(self, values: np.ndarray) -> float:
+        """Compute the Euclidean norm of a variable.
+
+        Parameters:
+            values: algebraic representation of a mixed-dimensional variable
+
+        Returns:
+            float: measure of values
+
+        """
+        return self._euclidean_norm(values)
+
+    def equation_norm(self, values: np.ndarray) -> float:
+        """Compute the Euclidean norm of an equation.
+
+        Parameters:
+            values: algebraic representation of a mixed-dimensional equation
+
+        Returns:
+            float: measure of values
+
+        """
+        return self._euclidean_norm(values)
+
+
+class MultiphysicsEuclideanMetric:
+    """Plain Euclidean norm for variables and equations, computed per variable and
+    equation block.
+
+    """
+
+    equation_system: pp.EquationSystem
+
+    def _euclidean_norm(self, values: np.ndarray) -> float:
+        """Compute the Euclidean norm of an array.
+
+        Parameters:
+            values: array to compute the norm of.
+
+        Returns:
+            float: measure of values
+
+        """
+        return np.linalg.norm(values) / np.sqrt(values.size)
+
+    def variable_norm(self, values: np.ndarray) -> dict[str, float]:
+        """Compute the Euclidean norm of each separate variable.
+
+        Parameters:
+            values: algebraic representation of a mixed-dimensional variable
+
+        Returns:
+            dict[str, float]: measure of values for each variable block
+
+        """
+        norms = {}
+        variable_blocks = {
+            variable.name: self.equation_system.dofs_of([variable])
+            for variable in self.equation_system.variables
+        }
+        for name, indices in variable_blocks.items():
+            norms[name] = self._euclidean_norm(values[indices])
+
+        return norms
+
+    def equation_norm(self, values: np.ndarray) -> dict[str, float]:
+        """Compute the Euclidean norm of each separate equation.
+
+        Parameters:
+            values: algebraic representation of a mixed-dimensional equation
+
+        Returns:
+            dict[str, float]: measure of values for each equation block
+
+        """
+        norms = {}
+        equation_blocks = self.equation_system.assembled_equation_indices
+        for name, indices in equation_blocks.items():
+            norms[name] = self._euclidean_norm(values[indices])
```

Comment:
We need to decide how to handle empty indices. Currently, this runs into division by zero.

Reply:
Not implemented yet: My suggestion would be to set these norms explicitly to zero, where indices empty.